sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala

```diff
@@ -327,11 +327,27 @@ private[joins] object UnsafeHashedRelation {
     // Create a mapping of buildKeys -> rows
     val keyGenerator = UnsafeProjection.create(key)
     var numFields = 0
+    val nullPaddingCombinations: Seq[UnsafeProjection] = if (isNullAware) {
+      // C(numKeys, 0), C(numKeys, 1) ... C(numKeys, numKeys - 1)
+      // In total 2^numKeys - 1 records will be appended.
+      key.indices.flatMap { n =>
+        key.indices.combinations(n).map { combination =>
+          // combination is Seq[Int] indicates which key should be replaced to null padding
+          val exprs = key.indices.map { index =>
+            if (combination.contains(index)) {
+              Literal.create(null, key(index).dataType)
```

Comment:
We need to expand it like this even if `key(index).nullable=false`?

Reply:
this is a good reminder, I have not considered on such case, i will confirm and reply you later.

Reply:
Since this works correctly,  I think it is okay to separate the optimization from this PR.

Reply:
It should be easy to fold this optimization into this PR ... All we need to do is to only create the combinations for the (truly) nullable keys. I think this is an important performance optimization since most keys are non nullable (in a well formed schema) and it would reduce the memory blow up.

Reply:
After serious consideration, I think when `key(index).nullable = false` case, it should be expand as well.
Just consider the following case, with streamedSide and buildSide both has only one record
(1, null, 3)   => (1, 2, 3)
this was consider a MATCH on NAAJ, right? no matter buildSide column is nullable or not, NAAJ care about its real key value of the record. if we don't expand (1,2,3) to 7X record, we can't find a exact match of (1,null,3)

@maropu @agrawaldevesh

Reply:
Makes sense and great observation. Worth explaining this in the comments and adding a test for this. 

I am wondering if you still have a test suite that compares the old-school BNLJ results vs with this optimization ? This would be a great test case.